src/compiler.ts

```diff
@@ -6366,6 +6406,28 @@ export class Compiler extends DiagnosticEmitter {
     }
   }
 
+  private injectClosedLocals(
+    closureContextLocal: Local
+  ): ExpressionRef {
+    var module = this.module;
+    var type = closureContextLocal.type;
+    var locals = type.locals!;
+    let _values = Map_values(locals);
+    var exprs = new Array<ExpressionRef>(_values.length);
+    for (let i = 0, k = _values.length; i < k; ++i) {
+      let local = unchecked(_values[i]);
+      let closureClass= type.classReference!
+      exprs[i] = module.store(
+        local.type.byteSize,
+        module.local_get(closureContextLocal.index, NativeType.I32),
+        module.local_get(local.index, local.type.toNativeType()),
+        local.type.toNativeType(),
+        closureClass.offsetof(local.name)
+      )
+    }
```

Comment:
```suggestion
    var locals = assert(type.locals);
    let _values = Map_values(locals);
    var exprs = new Array<ExpressionRef>(_values.length);
    assert(type.is(TypeFlags.REFERENCE));
    var closureClass = assert(type.classReference);
    for (let i = 0, k = _values.length; i < k; ++i) {
      let local = unchecked(_values[i]);
      let nativeType = local.type.toNativeType();
      exprs[i] = module.store(
        local.type.byteSize,
        module.local_get(closureContextLocal.index, NativeType.I32),
        module.local_get(local.index, nativeType),
        nativeType,
        closureClass.offsetof(local.name)
      )
    }
```

The difference of using an `assert` vs a `!` is that the former triggers an assertion when compiling the compiler to JS as well, not only when compiling to Wasm. I typically use `!` if I am 100% certain that something will never be false-ish anyway, and an `assert` if I'd like to perform an actual check.

Added the REFERENCE assert because a type can have a `classReference` but is not actually a class, theoretically, for instance it is planned that `i32` will eventually have the `I32` wrapper as its `classReference`, but remains a basic type.